src/stan/services/optimize/defaults.hpp

```diff
@@ -0,0 +1,326 @@
+          return "Convergence tolerance on changes in parameter value.";
+        }
+
+        /**
+         * Validates tol_param; tol_param must be greater than or equal to 0.
+         *
+         * @param[in] tol_param argument to validate
+         * @throw std::invalid argument if tol_param < 0
+         */
+        static void validate(double tol_param) {
+          if (!(tol_param >= 0))
+            throw std::invalid_argument("tol_param");
+        }
+
+        /**
+         * Default tol_param: 1e-08.
+         *
+         * @returns default
+         */
+        static double default_value() {
+          return 1e-08;
+        }
+      };
+
+      /**
+       * Amount of history to keep for L-BFGS.
+       */
+      struct history_size {
+        /**
+         * Description of history_size.
+         *
+         * @returns description
+         */
+        static std::string description() {
+          return "Amount of history to keep for L-BFGS.";
+        }
+
+        /**
+         * Validates history_size; history_size must be greater than 0.
+         *
+         * @param[in] history_size argument to validate
+         * @throw std::invalid argument if history_size <= 0
+         */
+        static void validate(int history_size) {
+          if (!(history_size > 0))
+            throw std::invalid_argument("history_size must be greater than 0.");
+        }
+
+        /**
+         * Default history_size: 5.
+         *
+         * @returns default
+         */
+        static int default_value() {
+          return 5;
+        }
+      };
+
+      /**
+       * Total number of iterations.
+       */
+      struct iter {
+        /**
+         * Description of iter.
+         *
+         * @returns description
+         */
+        static std::string description() {
+          return "Total number of iterations.";
+        }
+
+        /**
+         * Validates iter; iter must be greater than 0.
+         *
+         * @param[in] iter argument to validate
+         * @throw std::invalid argument if iter <= 0
+         */
+        static void validate(int iter) {
+          if (!(iter > 0))
+            throw std::invalid_argument("iter must be greater than 0.");
+        }
+
+        /**
+         * Default iter: 2000
+         *
+         * @returns default
+         */
+        static int default_value() {
+          return 2000;
+        }
+      };
+
+      /**
+       * Save optimization interations to output.
+       */
+      struct save_iterations {
+        /**
+         * Description of save_iterations.
+         *
+         * @returns description
+         */
+        static std::string description() {
+          return "Save optimization interations to output.";
+        }
+
+        /**
+         * Validates save_iterations. This is a noop.
+         *
+         * @param[in] save_iterations argument to validate
+         */
+        static void validate(bool save_iterations) {
```

Comment:
The validation happens from within the interfaces. Although these don't extend from a shared base class, the intent was to share a common concept that could be accessed programmatically from the interfaces. I think it's better to just leave the `validate()` functions there even if they are no-ops.